lib/utils/timer.js

```diff
@@ -16,29 +16,91 @@ const process = require('process');
 
 // Simple timer making process.hrtime() easy to use
 class Timer {
+    // creates a timer. use start() to start measurements
     constructor() {
-        this.id = process.hrtime();
+        this._measured = false;
+        this._current = 0.0;
+        this._total = 0.0;
     }
 
-    end() {
-        if (!this.hrtimeElapsed) {
+    // starts or continues the timer
+    start() {
+        if (!this._id) {
+            this._id = process.hrtime();
+
+            this._measured = true;
+        }
+        return this;
+    }
+
+    // stops the timer if it was running and returns the total time in seconds as float
+    stop() {
+        if (this._id) {
             try {
-                this.hrtimeElapsed = process.hrtime(this.id);
+                this._current = hrtimeToSeconds(process.hrtime(this._id));
+                this._total += this._current;
             } catch (e) {
                 console.error("error getting timing metrics:", e.message || e);
             }
+
+            delete this._id;
         }
-        return hrtimeToSeconds(this.hrtimeElapsed);
     }
 
-    // end() and duration() do the same, end the timer if not ended yet and return the duration
-    // but to the client it reads nicer depending on what they want
-    duration() {
-        return this.end();
+    // returns the duration between the last start() and stop().
+    // stops timer first if it is currently running.
+    // returns seconds as float. if the timer never ran, it returns undefined
+    currentDuration() {
+        this.stop();
+        return this._measured ? this._current : undefined;
+    }
+
+    // returns the total duration between all start() and stop() calls.
+    // stops timer first if it is currently running.
+    // returns seconds as float. if the timer never ran, it returns undefined
+    totalDuration() {
+        this.stop();
+        return this._measured ? this._total : undefined;
     }
 
-    durationSec() {
-        return this.end().toFixed(3);
+    // returns a human readable representation of the seconds with 3 decimal preceision
+    toString() {
+        this.stop();
+        return this._measured ? this._total.toFixed(3) : "???";
+    }
+
+    // returns a Timer with just the current duration
+    // useful for toString() for just current duration
+    current() {
+        const t = new Timer();
+        t._measured = true;
+        t._current = this._current;
+        t._total = this._current;
+        return t;
+    }
+
+    // sums multiple timer's currentDuration() and skips undefined values
+    static currentSum(...timers) {
+        let sum = 0.0;
+        timers.forEach(timer => {
+            const d = timer.currentDuration();
```

Comment:
This is hiding that the timers will be stopped (`currentDuration` calls stop). Is this the wanted behavior?

Reply:
Yes. `currentDuration()`, `totalDuration()` all stop() the timer.